### PR TITLE
Add uninstall subcommand

### PR DESCRIPTION
**Pull request by marcoesters**
### Description

Uninstalling distributions for `conda` currently only has incomplete solutions with files and broken auto-run/initializer scripts left behind (see https://github.com/conda/constructor/issues/642, https://github.com/conda/constructor/issues/588, and https://github.com/conda/constructor/issues/572). A clean uninstallation is currently not available without manually running a series of commands that are still not 100% safe - e.g., `conda init --reverse --all` will remove initialization of `conda`, no matter what installation it is pointed to.

This PR introduces a cross-platform uninstallation subcommand that uses conda-internal features to perform the following tasks:

* Uninstall all environments within a base environment or environments directory (such as a directory set by `envs_dirs`). This uninstalls all shortcuts, performs pre-unlink actions, and de-registers the environments from the `environments.txt` file.
* Runs `conda init reverse` if a configuration file points to any of the environments that are removed.
* Allows a user to remove `.condarc` files, either for the user, system-wide, or all.
* Allows the user to run `conda clean --all` to remove package caches outside the installation directory.
* Allows the user to remove cache directories, including the `.conda` file.

This is implemented in `conda-standalone` so that a single binary can be used on macOS and Linux. It also allows decoupling of the installer "front end" (built by `constructor`) from the installer "back end" (such as `conda-standalone` or `micromamba`), as outlined in the following issue: https://github.com/conda/constructor/issues/549

**Known gap**: Soft links are currently not well supported: files are unlinked, but no efforts are made to ensure that this doesn't leave unused files behind.

### Checklist - did you ...

- [X] Add a file to the `news` directory ([using the template](https://github.com/conda/conda-standalone/blob/main/news/TEMPLATE)) for the next release's release notes?
- [X] Add / update necessary tests?
- [X] Add / update outdated documentation?

### CODE REVIEW

File: src/entry_point.py [hunk 1]
```diff
@@ -315,6 +662,9 @@ def main():
             return _constructor_subcommand()
         elif sys.argv[1] == "python":
             return _python_subcommand()
+        elif sys.argv[1] == "uninstall":
```
[the remark this answers is not in the document]
Comment (marcoesters) [reply src/entry_point.py]: Making it a subcommand works, but comes with drawbacks. I had to make the mutually exclusive group and the `--prefix` argument optional.

I can easily revert this though and instead rename the subcommand (e.g., "uninstall-prefix").

[the remark this answers is not in the document]
Comment (marcoesters) [reply src/entry_point.py]: How would you feel about `--prefix` defaulting to `str(Path(sys.executable).parent)`?

[the remark this answers is not in the document]
Comment (jaimergp) [reply src/entry_point.py]: Isn't the parent of sys.executable for `conda-standalone` the temporary extraction directory?

File: news/112-add-uninstall-subcommand [hunk 2]
```diff
@@ -0,0 +1,20 @@
+### Enhancements
+
+* Add `uninstall` subcommand, which uninstalls all environments inside a prefix and reverses
+  `conda init` commands. It also for deleting chache directories and configuration files. (#112)
```
Comment (jaimergp) [inline news/112-add-uninstall-subcommand:4]:
```suggestion
  `conda init` commands. It also for deleting cache directories and configuration files. (#112)
```

File: src/entry_point.py [hunk 3]
```diff
@@ -136,9 +142,58 @@ def __call__(self, parser, namespace, values, option_string=None):
...
+            " Not recommended when multiple conda installations are on the system"
+            " or when running on an environments directory."
+        ),
+    )
+
     args, args_unknown = p.parse_known_args()
 
-    args.prefix = os.path.abspath(args.prefix)
+    if args.prefix is None:
+        raise argparse.ArgumentError("the following arguments are required: --prefix")
```
Comment (jaimergp) [inline src/entry_point.py:194]: Yep this was going to be my workaround. 👍

File: src/entry_point.py [hunk 4]
```diff
@@ -201,6 +256,311 @@ def _constructor_menuinst(prefix, pkg_names=None, root_prefix=None, remove=False
         install(str(json_path), remove=remove, prefix=prefix, root_prefix=root_prefix)
 
 
+def _is_subdir(directory: Path, root: Path) -> bool:
```
Comment (jaimergp) [inline src/entry_point.py:259]: Initially I thought this would be something about `CONDA_SUBDIR`, but it's not, right?

File: src/entry_point.py [hunk 5]
```diff
@@ -201,6 +256,311 @@ def _constructor_menuinst(prefix, pkg_names=None, root_prefix=None, remove=False
...
+    """
+    import re
+
+    from conda.base.constants import COMPATIBLE_SHELLS
+    from conda.common.compat import on_win
+    from conda.common.path import win_path_to_unix
+    from conda.core.initialize import (
+        CONDA_INITIALIZE_PS_RE_BLOCK,
+        CONDA_INITIALIZE_RE_BLOCK,
+        _read_windows_registry,
```
Comment (jaimergp) [inline src/entry_point.py:290]: Hm, this is going to be a bit brittle... I'm not sure if `conda` respects their deprecation policy for `_private` names. Do they?

File: src/entry_point.py [hunk 6]
```diff
@@ -201,6 +256,311 @@ def _constructor_menuinst(prefix, pkg_names=None, root_prefix=None, remove=False
...
+                    # Ignore .exe suffix to make the logic simpler
+                    if shell in ("csh", "tcsh") and sys.platform != "win32":
+                        sentinel_str = str(prefix / "etc" / "profile.d" / "conda.csh")
+                    else:
+                        sentinel_str = str(prefix / BIN_DIRECTORY / "conda")
+                    if sys.platform == "win32" and shell != "powershell":
+                        sentinel_str = win_path_to_unix(sentinel_str)
+                        # Remove /cygdrive to make the path shell-independent
+                        if sentinel_str.startswith("/cygdrive"):
+                            sentinel_str = sentinel_str[9:]
```
Comment (jaimergp) [inline src/entry_point.py:347]: You can use https://docs.python.org/3/library/stdtypes.html#str.removeprefix now :D

File: src/entry_point.py [hunk 7]
```diff
@@ -201,6 +256,311 @@ def _constructor_menuinst(prefix, pkg_names=None, root_prefix=None, remove=False
...
+            if not file.exists():
+                return
+            if file.is_symlink() or file.is_file():
+                file.unlink()
+            elif file.is_dir():
+                rmtree(file)
+        except PermissionError:
+            pass
+
+    def _remove_file_and_parents(file: Path, stop_at: Path | None = None):
```
Comment (jaimergp) [inline src/entry_point.py:401]: This can be quite a destructive function, so I'd recommend adding separate tests and making it a global function.

For example, we need to ensure that `file` is contained in `stop_at`, right?

File: tests/test_uninstall.py [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+    run_plan,
+    run_plan_elevated,
+)
+from conftest import _get_shortcut_dirs, menuinst_pkg_specs, run_conda
+from ruamel.yaml import YAML
+
+ON_WIN = sys.platform == "win32"
+ON_MAC = sys.platform == "darwin"
+ON_LINUX = not (ON_WIN or ON_MAC)
+ON_CI = bool(os.environ.get("CI"))
```
Comment (jaimergp) [inline tests/test_uninstall.py:22]: This will be true for `CI=0` 🤔

File: tests/test_uninstall.py [hunk 9]
```diff
@@ -0,0 +1,405 @@
...
+
+    monkeypatch.setenv("XDG_CONFIG_HOME", str(paths["confighome"]))
+    monkeypatch.setenv("XDG_CACHE_HOME", str(paths["cachehome"]))
+    monkeypatch.setenv("XDG_DATA_HOME", str(paths["datahome"]))
+    monkeypatch.setenv("BINSTAR_CONFIG_DIR", str(paths["binstar"]))
+
+    return paths
+
+
+def create_env(
```
Comment (jaimergp) [inline tests/test_uninstall.py:66]: I guess we are not using stuff from `conda.testing` like `tmp_env` because we want to handle the removal ourselves, right?

File: src/entry_point.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (marcoesters) [reply src/entry_point.py]: No, `sys.executable` is the path to `conda.exe`, so the default would be the directory `conda.exe` is in.

`prefix` and similar variables point to the extraction directory though.

File: src/entry_point.py (hunk 4 above)
Comment (marcoesters) [reply src/entry_point.py:259]: Correct, it does not refer to `CONDA_SUBDIR`, but to subdirectories on the file system.

I created this helper function because I need it in a few places during the uninstallation process.

File: src/entry_point.py (hunk 5 above)
Comment (marcoesters) [reply src/entry_point.py:290]: I'm not sure, but if `conda` changes this, `test_uninstallation_init_reverse` will fail and we can change the name of the function.

The alternative is to copy and paste the code. However, I want to re-use as many `conda` components as possible to be closer to `conda` behavior.

File: src/entry_point.py [hunk 10]
```diff
@@ -201,6 +256,311 @@ def _constructor_menuinst(prefix, pkg_names=None, root_prefix=None, remove=False
...
+):
+    """
+    Remove a conda prefix or a directory containing conda environments.
+
+    This command also provides options to remove various cache and configuration
+    files to fully remove a conda installation.
+    """
+    from conda.base.constants import PREFIX_MAGIC_FILE
+
+    ENVS_DIR_MAGIC_FILE = ".conda_envs_dir_test"
```
[the remark this answers is not in the document]
Comment (marcoesters) [reply src/entry_point.py:369]: This is part of conda. I added a comment with the link to it.

While the comment says this shouldn't be relied upon, this is the magic file and has been for six years. I think we can consider this stable. If it changes, `test_uninstallation_envs_dirs` will fail and we can adjust.

File: src/entry_point.py (hunk 7 above)
Comment (marcoesters) [reply src/entry_point.py:402]: I added a check, which actually made me find a bug. This made me add a warning to the README file that support for softlinks is limited.

The problem with writing separate tests is that we would have to create an entry point for it, which would expose that destructive function even more. I added another tests to ensure that the function does not a non-empty directory.

File: tests/test_uninstall.py (hunk 8 above)
Comment (marcoesters) [reply tests/test_uninstall.py]: True, I improved that check. We have a similar thing in `constructor` - I will port it there, too.

File: tests/test_uninstall.py (hunk 9 above)
Comment (marcoesters) [reply tests/test_uninstall.py]: There is a much simpler reason: I didn't have `conda` as a test dependency when I wrote that function and it was too small of a function for me to import `conda`.

However, since I need it for other things now, I replaced almost instances with the conda fixture. I had to keep `run_conda` for `menuinst` though because I cannot monkeypatch the paths otherwise.

File: tests/test_uninstall.py [hunk 11]
```diff
@@ -0,0 +1,405 @@
...
+
+@pytest.fixture(scope="function")
+def mock_system_paths(
+    monkeypatch,
+    tmp_path: Path,
+) -> dict[str, Path]:
+    paths = {}
+    if ON_WIN:
+        homedir = tmp_path / "Users" / "user"
+        monkeypatch.setenv("USERPROFILE", str(homedir))
```
[the remark this answers is not in the document]
Comment (marcoesters) [reply tests/test_uninstall.py:44]: Can't hurt

File: src/entry_point.py [hunk 12]
```diff
@@ -85,11 +85,17 @@ def __call__(self, parser, namespace, values, option_string=None):
...
             setattr(namespace, self.dest, num)
 
-    p = argparse.ArgumentParser(description="constructor helper subcommand")
+    # Remove "constructor" so that it does not clash with the uninstall subcommand
+    del sys.argv[1]
+    p = argparse.ArgumentParser(
+        prog="conda.exe constructor", description="constructor helper subcommand"
+    )
+    # Cannot make this a required argument or `conda.exe constructor uninstall --prefix`
+    # will not work (would have to be `conda constructor --prefix uninstall`)
```
Comment (jaimergp) [inline src/entry_point.py:94]: We are checking against `prefix=None` later right? So in practice it is required for the user. A comment here to clarify would be perfect.

File: src/entry_point.py [hunk 13]
```diff
@@ -110,7 +116,7 @@ def __call__(self, parser, namespace, values, option_string=None):
         f"Defaults to {DEFAULT_NUM_PROCESSORS}.",
     )
 
-    g = p.add_mutually_exclusive_group(required=True)
+    g = p.add_mutually_exclusive_group()
```
Comment (jaimergp) [inline src/entry_point.py:119]: We'd need to add checks to make this required manually if `uninstall` is not in the ARGV.

File: src/entry_point.py (hunk 5 above)
Comment (jaimergp) [reply src/entry_point.py:290]: Ok, good point, we'll catch this at build time because it's frozen anyway.

File: src/entry_point.py [hunk 14]
```diff
@@ -201,6 +256,314 @@ def _constructor_menuinst(prefix, pkg_names=None, root_prefix=None, remove=False
...
+
+    This command also provides options to remove various cache and configuration
+    files to fully remove a conda installation.
+    """
+    from conda.base.constants import PREFIX_MAGIC_FILE
+
+    # See: https://github.com/conda/conda/blob/475e6acbdc98122fcbef4733eb8cb8689324c1c8/conda/gateways/disk/create.py#L482-L488 # noqa
+    ENVS_DIR_MAGIC_FILE = ".conda_envs_dir_test"
+
+    uninstall_prefix = Path(uninstall_dir).expanduser().resolve()
```
Comment (jaimergp) [inline src/entry_point.py:371]: Do we need to `os.path.expandvars` too or is that not expected?

File: src/entry_point.py (hunk 7 above)
Comment (jaimergp) [reply src/entry_point.py:402]: Hm, cannot we import the `entry_point.py` module in the tests directly and target that function?

File: src/entry_point.py [hunk 15]
```diff
@@ -201,6 +256,314 @@ def _constructor_menuinst(prefix, pkg_names=None, root_prefix=None, remove=False
...
+            return
+        _remove_file_directory(file)
+        parent = file.parent
+        while (
+            parent != stop_at and parent.is_dir() and not next(parent.iterdir(), None)
+        ):
+            _remove_file_directory(parent)
+            parent = parent.parent
+
+    def _remove_config_file_and_parents(file: Path):
```
Comment (jaimergp) [inline src/entry_point.py:418]: There's precedent for users not liking this default behaviour, see https://github.com/conda/conda/issues/13113. Not sure if we should delete empty parents recursively 🤔  Maybe just the config-containing one? Happy to have my mind changed here, though, so let me know.

File: src/entry_point.py [hunk 16]
```diff
@@ -201,6 +256,314 @@ def _constructor_menuinst(prefix, pkg_names=None, root_prefix=None, remove=False
...
+
+    print(f"Uninstalling conda installation in {uninstall_prefix}...")
+    prefixes = [
+        file.parent.parent.resolve()
+        for file in uninstall_prefix.glob(f"**/{PREFIX_MAGIC_FILE}")
+    ]
+    # Sort by path depth. This will place the root prefix first
+    # Since it is more likely that profiles contain the root prefix,
+    # this makes loops more efficient.
+    prefixes.sort(key=lambda x: len(x.parts))
```
Comment (jaimergp) [inline src/entry_point.py:448]: 👏 👏 👏

File: README.md [hunk 17]
```diff
@@ -50,21 +50,64 @@ It also adds new subcommands not available on the regular `conda`:
...
+                     Not recommended with multiple conda installations when softlinks are enabled.
+
+> [!IMPORTANT]
+> Use `sudo -E` if removing system-level configuration files requires superuser privileges.
+> `conda` relies on environment variables like `HOME` and `XDG_CONFIG_HOME` when detecting
+> configuration files, which may be overwritten with just `sudo`.
+> This can cause files to be left behind.
+
+> [!WARNING]
+> Support for softlinks is still limited.
```
Comment (jaimergp) [inline README.md:109]: Can you elaborate? Are errors expected, or undefined behavior?

File: src/entry_point.py [hunk 18]
```diff
@@ -136,9 +142,58 @@ def __call__(self, parser, namespace, values, option_string=None):
...
+        help=(
+            "Remove all .condarc files."
+            " `user` removes the files inside the current user's"
+            " home directory and `system` removes all files outside of that directory."
+            " Not recommended when multiple conda installations are on the system"
+            " or when running on an environments directory."
+        ),
+    )
+    uninstall_subcommand.add_argument(
+        "--remove-caches",
```
Comment (jaimergp) [inline src/entry_point.py:180]:
```suggestion
        "--remove-additional-caches",
```

The term is a bit convoluted with `pkgs/` and `pkgs/cache` being taken care of in `conda-clean`, but this argument targets `~/.conda` and `anaconda-client` stuff.

File: src/entry_point.py [hunk 19]
```diff
@@ -136,9 +142,58 @@ def __call__(self, parser, namespace, values, option_string=None):
...
+        action="store_true",
+        required=False,
+        help=(
+            "Remove all cache directories created by conda."
+            " This includes the .conda directory inside HOME/USERPROFILE."
```
Comment (jaimergp) [inline src/entry_point.py:185]:
```suggestion
            " This includes the ~/.conda directory and anaconda-client data."
```

File: src/entry_point.py [hunk 20]
```diff
@@ -201,6 +256,314 @@ def _constructor_menuinst(prefix, pkg_names=None, root_prefix=None, remove=False
...
+        for_system = not for_user
+        anaconda_prompt = False
+        plan = _get_init_reverse_plan(
+            uninstall_prefix, prefixes, for_user, for_system, anaconda_prompt
+        )
+        # Do not call conda.core.initialize() because it will always run make_install_plan.
+        # That function will search for activation scripts in sys.prefix which do no exist
+        # in the extraction directory of conda-standalone.
+        run_plan(plan)
+        run_plan_elevated(plan)
```
[the remark this answers is not in the document]
Comment (marcoesters) [reply src/entry_point.py:462]: As far as I know, that's correct. There doesn't seem to be a "failed" state in the `Results` enum either.

I am printing the plan results at least so that users can check if they don't see the results they expect.

File: src/entry_point.py (hunk 15 above)
Comment (marcoesters) [reply src/entry_point.py:412]: That's what I was intending with `rootdir`, but it was still a little too aggressive. I changed the selection of which directories to remove. The only thing that might be controversial is the `~/.config` directory. I think it's okay to delete if empty. I would at least argue that we should delete it on Windows where configs are stored in different places.

File: src/entry_point.py (hunk 14 above)
Comment (marcoesters) [reply src/entry_point.py]: Good call. I added it to where we parse `args.prefix`.

File: src/entry_point.py (hunk 13 above)
Comment (marcoesters) [reply src/entry_point.py:121]: I added that into the same function and created tests to check.

File: src/entry_point.py (hunk 7 above)
Comment (marcoesters) [reply src/entry_point.py]: This function is obsolete now that I don't aggressively remove parents anymore. So, no tests needed!

File: src/entry_point.py [hunk 21]
```diff
@@ -201,6 +256,314 @@ def _constructor_menuinst(prefix, pkg_names=None, root_prefix=None, remove=False
...
+    if conda_clean:
+        conda_main("clean", "--all", "-y")
+        # Delete empty package cache directories
+        for directory in context.pkgs_dirs:
+            pkgs_dir = Path(directory)
+            if not pkgs_dir.exists():
+                continue
+            expected_files = [pkgs_dir / "urls", pkgs_dir / "urls.txt"]
+            if all(file in expected_files for file in pkgs_dir.iterdir()):
+                _remove_file_and_parents(pkgs_dir)
```
[the remark this answers is not in the document]
Comment (marcoesters) [reply src/entry_point.py]: I can get on board for the package cache, but for the binstar directory and the notices cache, we should remove the parents:

* The anaconda-client token is either in a `$BINSTAR_TOKEN_DIR/data` or a `binstar/ContinuumIO` (Windows) or `binstart` directory and is created by `conda` via `anaconda-client`. I think we can remove the binstar directories at least. To be honest, the best solution would be to make `anaconda-client` a plug-in handle that directory when running `conda remove`. Right now, however, `conda` is the API that's creating that directory, so I decided to handle it here.
* The notices cache is written into a specific conda cache directory, so we will not remove a user-created directory. Unfortunately, the subdirectories are different for Windows and Unix, but they are all called `conda`. So, I changed it to `_remove_config_file_and_parents` here.

File: README.md (hunk 17 above)
Comment (marcoesters) [reply README.md]: We only unlink softlinks and do not do anything with the underlying files. I made that more explicit.

File: src/entry_point.py (hunk 18 above)
Comment (marcoesters) [reply src/entry_point.py]: Yes, unfortunately, `conda` overloads the term "cache". I changed it to `--remove-conda-caches` because it's more descriptive than `--remove-additional-caches`.

Comment (jaimergp) [reply src/entry_point.py]: I still think that if a user sees both `--conda-clean` and `--remove-conda-caches` might get confused about the difference between both. Maybe the best option is to split it explicitly: `--remove-anaconda-client-data`, `--remove-conda-notices-data`. I'd argue the second one belongs in `conda clean` responsibilities though, so what iffff:

- The `--conda-clean` flag _also_ takes care of conda notices, there's no choice. Eventually we upstream this to `conda`.
- Folks can optionally remove `anaconda-client` data.

File: src/entry_point.py [hunk 22]
```diff
@@ -368,7 +368,7 @@ def _constructor_uninstall_subcommand(
     # See: https://github.com/conda/conda/blob/475e6acbdc98122fcbef4733eb8cb8689324c1c8/conda/gateways/disk/create.py#L482-L488 # noqa
     ENVS_DIR_MAGIC_FILE = ".conda_envs_dir_test"
 
-    uninstall_prefix = Path(uninstall_dir).expanduser().resolve()
+    uninstall_prefix = Path(uninstall_dir).resolve()
```
[the remark this answers is not in the document]
Comment (marcoesters) [reply src/entry_point.py:381]: Not quite - `.resolve()` also resolves links. However, I'm okay with removing it until we have a better plan for how to handle symlinks.

File: src/entry_point.py (hunk 18 above)
Comment (marcoesters) [reply src/entry_point.py]: I'm more amenable to the option that `--clean-caches` removes the notices cache and runs `conda clean` and that `--clean-conda-data` or `--clean-user-data` removes `~/.conda` and `anaconda-client` data.

Comment (marcoesters) [reply src/entry_point.py]: The more I think about this (and after talking to a conda maintainer) the more convinced I am that we shouldn't handle `anaconda-client` data here. This should be handled by the package instead. Otherwise, we would have to implement clean-up operations for other packages or an arbitrary selection.

I think `conda-standalone` should only handle `conda` files. So, my final proposal is:

* `--clean-caches` to run `conda clean` and remove the notices cache.
* `--remove-user-data` to just delete `~/.conda` for now.

This also follows somewhat the `XDG_*` convention where we have cache, data, and configs (here `.condarc` files). In that case, I can also rename `--remove-condarcs` to `--remove-config-files` to allow `constructor` to use more generic terms.

What do you think about this solution?

File: src/entry_point.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (jaimergp) [reply src/entry_point.py]: Ah, sorry, mixed it with `realpath`. nvm then!

File: src/entry_point.py (hunk 18 above)
Comment (jaimergp) [reply src/entry_point.py]: Love this last proposal and the XDG rationale. I think we should stick to a verb only: either `clean`, `clear` or `remove`, but not a mix. So pick one (using `clean` as an example) and then we say:

- `--clean-caches`
- `--clean-user-data`
- `--clean-config-files`

Comment (marcoesters) [reply src/entry_point.py]: I prefer `remove-*` because it tells the user that files will be deleted.

File: README.md [hunk 23]
```diff
@@ -50,21 +50,65 @@ It also adds new subcommands not available on the regular `conda`:
...
+There are several options to remove configuration and cache files:
+
+```bash
+$ conda.exe constructor uninstall [-h] --prefix PREFIX [--conda-clean] [--remove-config-files {user,system,all}]
+                                  [--remove-conda-caches]
+```
+
+- `--prefix` (required): Path to the conda directory to uninstall.
+- `--remove-caches`:
+  Removes the notices cache and runs conda --clean --all to clean package caches outside the
```
Comment (jaimergp) [inline README.md:91]:
```suggestion
  Removes the notices cache and runs `conda --clean --all` to clean package caches outside the
```

File: README.md [hunk 24]
```diff
@@ -50,21 +50,65 @@ It also adds new subcommands not available on the regular `conda`:
...
+  installation directory. This is especially useful when `pkgs_dirs` is set in a `.condarc` file.
+  Not recommended with multiple conda installations when softlinks are enabled.
+- `--remove-config-files {user,system,all}`:
+   Removes all .condarc files. `user` removes the files inside the current user's home directory
```
Comment (jaimergp) [inline README.md:95]:
```suggestion
   Removes all `.condarc` files. `user` removes the files inside the current user's home directory
```

File: README.md [hunk 25]
```diff
@@ -50,21 +50,65 @@ It also adds new subcommands not available on the regular `conda`:
...
+   and `system` removes all files outside of that directory. Not recommended when multiple conda
+   installations are on the system or when running on an environments directory.
+- `--remove-user-data`:
+  Removes the ~/.conda directory. Not recommended when multiple conda installations are installed
```
Comment (jaimergp) [inline README.md:99]:
```suggestion
  Removes the `~/.conda` directory. Not recommended when multiple conda installations are installed
```

File: src/entry_point.py [hunk 26]
```diff
@@ -136,9 +144,66 @@ def __call__(self, parser, namespace, values, option_string=None):
...
+            " or when running on an environments directory."
+        ),
+    )
+
     args, args_unknown = p.parse_known_args()
 
-    args.prefix = os.path.abspath(args.prefix)
+    if args.command != "uninstall":
+        group_args = getattr(g, "_group_actions")
+        if all(getattr(args, arg.dest) is None for arg in group_args):
```
Comment (jaimergp) [inline src/entry_point.py:197]:
```suggestion
        if all(getattr(args, arg.dest, None) is None for arg in group_args):
```

default param missing?

File: src/entry_point.py [hunk 27]
```diff
@@ -136,9 +144,66 @@ def __call__(self, parser, namespace, values, option_string=None):
...
+            required_args = [arg.option_strings[0] for arg in group_args]
+            raise argparse.ArgumentError(
+                f"one of the following arguments are required: {'/'.join(required_args)}"
+            )
+
+    if args.prefix is None:
+        raise argparse.ArgumentError("the following arguments are required: --prefix")
+
+    args.prefix = os.path.abspath(os.path.expanduser(os.path.expandvars(args.prefix)))
     args.root_prefix = os.path.abspath(os.environ.get("CONDA_ROOT_PREFIX", args.prefix))
```
Comment (jaimergp) [inline src/entry_point.py:207]: I guess we don't need to expand {user,vars} `root_prefix` too, or do we?

File: tests/test_uninstall.py [hunk 28]
```diff
@@ -0,0 +1,432 @@
...
+    remove_system = remove != "user"
+    remove_user = remove != "system"
+    condarc = {
+        "channels": [CONDA_CHANNEL],
+    }
+    needs_sudo = False
+    if ON_WIN:
+        system_condarc = Path("C:/ProgramData/conda/.condarc")
+    else:
+        system_condarc = Path("/etc/conda/.condarc")
```
Comment (jaimergp) [inline tests/test_uninstall.py:375]: Should we rely on env vars like `PROGRAMDATA` and `XDG_DATA_*` for this test? Or do you think it's not necessary?

Comment (marcoesters) [reply tests/test_uninstall.py:375]: I 100% agree that we should rely on `PROGRAMDATA`, but unfortunately, `conda` doesn't use those variables: https://github.com/conda/conda/blob/main/conda/base/constants.py#L29-L42

File: src/entry_point.py (hunk 27 above)
Comment (marcoesters) [reply src/entry_point.py:207]: I'm not sure how `os.environ` expands variables here. I'd assume that they are already expanded?

**Review by jaimergp — APPROVED** [review]
Awesome work! Thanks for making it all the way through the leeengthy review!
